### PR TITLE
control-service: asynchronous deployment deletion

### PR DESCRIPTION
**Pull request by mivanov1988**
Why:
As part of VEP-2272, we need to introduce a process for synchronizing data jobs from the database to Kubernetes. Currently, the process lacks the ability to delete deployments asynchronously.

What:
We have introduced an asynchronous method for deployment deletion. It is based on the logic that if the desired deployment is null and the actual deployment is not null, the process performs deployment deletion.

Testing done:
Integration tests.

Signed-off-by: Miroslav Ivanov miroslavi@vmware.com

### CODE REVIEW

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/DataJobsDeploymentController.java [hunk 1]
```diff
@@ -51,14 +53,27 @@ public class DataJobsDeploymentController implements DataJobsDeploymentApi {
...
+
   @Override
   public ResponseEntity<Void> deploymentDelete(
       String teamName, String jobName, String deploymentId) {
     if (jobsService.jobWithTeamExists(jobName, teamName)) {
       // TODO: deploymentId not implemented
       if (jobName != null) {
-        deploymentService.deleteDeployment(jobName);
-        return ResponseEntity.accepted().build();
+        if (dataJobDeploymentPropertiesConfig
```
Comment (antoniivanov) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/DataJobsDeploymentController.java:66]: I feel it's better this `if dataJobDeploymentPropertiesConfig.writeTo` it's in one place and not so many. 

And the appropriate DeploymentService is injected depending on dataJobDeploymentPropertiesConfig.writeTo value. 

So that may necessitate having DeploymentServiceInterface or something but that would be more maintainable, readable and easier to remove in the future if/when we decide to clean up.

Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/DataJobsDeploymentController.java:66]: I am not sure how much easier it would be if we introduced an interface, the factory, and then, when we decide to clean up the code, reverted back to the same injection strategy. If you have a strong opinion I will refactor the code.

**Review by dakodakov — APPROVED** [review]
LGTM

**Review by antoniivanov — COMMENTED** [review]
What happens if data job is deleted?

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/DataJobsDeploymentController.java (hunk 1 above)
Comment (antoniivanov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/datajobs/DataJobsDeploymentController.java:66]: ok

**Comment by mivanov1988** [comment]
> What happens if data job is deleted?

That is a very good point! When you delete a data job, the database will automatically remove the associated deployments from the desired and actual tables. Additionally, the old DeploymentService.deleteDeployment() will be called, and the K8s artifacts will be deleted as well. However, because the synchronization process caches the data jobs and deployments, if a user deletes a data job during synchronization, the Kubernetes artifacts will be left in the cluster, potentially leading to an inconsistent state.

In my opinion, we should remove the references from the desired and actual deployment tables to the data job table. In this scenario, we can delete the data job record from the 'data_job' table and the deployment record from the 'desired_data_job_deployment' table. Subsequently, we will have an existing deployment record in the 'actual_data_job_deployment' table, which will be deleted by the synchronization process during the next synchronization, along with all K8S-related artifacts.

@antoniivanov what do you think?

**Comment by mivanov1988** [comment]
> > > What happens if data job is deleted?
> > 
> > 
> > That is a very good point! When you delete a data job, the database will automatically remove the associated deployments from the desired and actual tables. Additionally, the old DeploymentService.deleteDeployment() will be called, and the K8s artifacts will be deleted as well. However, because the synchronization process caches the data jobs and deployments, if a user deletes a data job during synchronization, the Kubernetes artifacts will be left in the cluster, potentially leading to an inconsistent state.
> > In my opinion, we should remove the references from the desired and actual deployment tables to the data job table. In this scenario, we can delete the data job record from the 'data_job' table and the deployment record from the 'desired_data_job_deployment' table. Subsequently, we will have an existing deployment record in the 'actual_data_job_deployment' table, which will be deleted by the synchronization process during the next synchronization, along with all K8S-related artifacts.
> > @antoniivanov what do you think?
> 
> Seems a good idea. It make sense that user should not be able to directly touch actual_deployment , this should be done only by the automation. Users (aka API endpoints) interact only with desired state
> 
> We need to make sure to test some corner cases:
> 
> * synchronization process is delayed or fails  (pod restarts) for some reason after the data job is deleted
It will delete the cron job on the next synchronization.
> * job is created almost immediately with the same name after starting deletion
It is okay. The cron job will be deleted, and the job will no longer have an associated one. The UI might display the existing deployment for a short period of time.
> * What would the GraphQL endpoint return (aka the UI show) Probably the job as deleted and hence nothing
Nothing happens because the job is a first-class citizen.

File: projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/service/deploy/DataJobDeploymentCrudITV2.java [hunk 2]
```diff
@@ -159,6 +159,12 @@ public void testSynchronizeDataJob() throws Exception {
     Assertions.assertNotEquals(lastDeployedDateInitial, lastDeployedDateShouldBeChanged);
     Assertions.assertNotEquals(
         deploymentVersionShaShouldNotBeChanged, deploymentVersionShaShouldBeChanged);
+
+    // Deletes deployment
```
Comment (antoniivanov) [inline projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/service/deploy/DataJobDeploymentCrudITV2.java:163]: We should really considering refactoring those long test methods into multiple methods and using @Order annotation to dictate the order . 

It's not something you need to do now, but you can add a TODO if you want for next developer to consider doing this.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java [hunk 3]
```diff
@@ -144,6 +144,8 @@ void synchronizeDataJob(
           desiredDataJobDeployment,
           actualDataJobDeployment,
           isDeploymentPresentInKubernetes);
+    } else if (actualDataJobDeployment != null) {
```
Comment (antoniivanov) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java:147]: Where are the tests that verify `synchronizeDataJobs` (all jobs) work as expected? I see tests that `synchronizeDataJob` (singular) works. But not for the full sync.

I am suspicious deleting job might not work because we are iterating using `dataJobsFromDB` but if we delete a data job then `dataJobsFromDB` would not contain our deleted job.

File: projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/service/deploy/DataJobDeploymentCrudITV2.java (hunk 2 above)
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/service/deploy/DataJobDeploymentCrudITV2.java:163]: Yes, this is planned.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java (hunk 3 above)
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java:147]: Correct, the synchronizeDataJobs does not cover this case.

Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java:147]: Are you ok to implement this in a separate PR?

Comment (antoniivanov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DataJobsSynchronizer.java:147]: Yeah.

**Review by antoniivanov — APPROVED** [review]
